pulsar-broker/src/main/java/org/apache/pulsar/broker/loadbalance/extensions/channel/ServiceUnitStateChannelImpl.java

```diff
@@ -1505,7 +1502,21 @@ private void doHealthCheckBrokerAsyncWithRetries(String brokerId, int retry, Com
         }
     }
 
-    private synchronized void doCleanup(String broker, boolean gracefully) {
+    private void tryWaitForOverrides(List<CompletableFuture<Void>> overrideFutures, boolean force) {
```

Comment:
It seems that this solution is fine for this use case, but it's worth knowing that there's a generic solution for limiting concurrency for async operations which is used in Pulsar Admin client (introduced by PR https://github.com/apache/pulsar/pull/22541). It's the [com.spotify.futures.ConcurrencyReducer class](https://github.com/spotify/completable-futures/blob/master/src/main/java/com/spotify/futures/ConcurrencyReducer.java) from Spotify completable-futures library.

Reply:
Thank you for sharing this class. I think this simple array list is fine for this case as requests are added in a synchronized manner by the single load manager executor thread.